kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/GCSToBQLoadRunnable.java

```diff
@@ -0,0 +1,232 @@
+      logger.error("Found blob {}\\{} with no metadata.", blob.getBucket(), blob.getName());
+      return null;
+    }
+
+    String serializedTableId = blob.getMetadata().get(GCSToBQWriter.GCS_METADATA_TABLE_KEY);
+    Matcher matcher = METADATA_TABLE_PATTERN.matcher(serializedTableId);
+    String project = matcher.group("project");
+    String dataset = matcher.group("dataset");
+    String table = matcher.group("table");
+
+    if (project == null || dataset == null || table == null) {
+      logger.error("Found blob `{}/{}` without parsable table metadata.",
+                   blob.getBucket(), blob.getName());
+      return null;
+    }
+
+    return TableId.of(project, dataset, table);
+  }
+
+  private Map<Job, List<Blob>> triggerBigQueryLoadJobs(Map<TableId, List<Blob>> tablesToBlobs) {
+    Map<Job, List<Blob>> newJobs = new HashMap<>(tablesToBlobs.size());
+    for (Map.Entry<TableId, List<Blob>> entry : tablesToBlobs.entrySet()) {
+      newJobs.put(triggerBigQueryLoadJob(entry.getKey(), entry.getValue()), entry.getValue());
+    }
+    return newJobs;
+  }
+
+  private Job triggerBigQueryLoadJob(TableId table, List<Blob> blobs) {
+    List<String> uris = blobsToURIs(blobs);
+    // create job load configuration
+    LoadJobConfiguration loadJobConfiguration =
+        LoadJobConfiguration.newBuilder(table, uris)
+            .setFormatOptions(FormatOptions.json())
+            .setCreateDisposition(JobInfo.CreateDisposition.CREATE_IF_NEEDED)
+            .build();
+    // create and return the job.
+    Job job = bigQuery.create(JobInfo.of(loadJobConfiguration));
+    // update active jobs and claimed blobs.
+    activeJobs.put(job, blobs);
+    claimedBlobs.addAll(blobs);
+    logger.info("Triggered load job for table {} with {} blobs.", table, blobs.size());
+    return job;
+  }
+
+  private List<String> blobsToURIs(List<Blob> blobs) {
+    List<String> uris = new ArrayList<>(blobs.size());
+    for (Blob blob: blobs) {
+      uris.add(String.format(SOURCE_URI_FORMAT, bucket.getName(), blob.getName()));
+    }
+    return uris;
+  }
+
+  /**
+   * Check all active jobs. Remove those that have completed successfully and log a message for
+   * any jobs that failed. We only log a message for failed jobs because those blobs will be
+   * retried during the next run.
+   */
+  private void checkJobs() {
+    if (activeJobs.isEmpty()) {
+      // quick exit if nothing needs to be done.
+      return;
+    }
+    Iterator<Job> jobIterator = activeJobs.keySet().iterator();
+    int successCount = 0;
+    int failureCount = 0;
+    int blobsDeleted = 0;
+    while (jobIterator.hasNext()) {
+      Job job = jobIterator.next();
+      try {
+        if (job.isDone()) {
+          List<Blob> blobsToDelete = activeJobs.remove(job);
+          successCount++;
+          for (Blob blob : blobsToDelete) {
+            try {
+              blob.delete();
+              claimedBlobs.remove(blob);
```

Comment:
Wondering a little bit about failure scenarios. If a thread fails for some reason (e.g. delete() fails), activeJobs would be mutated but claimedBlobs would not. Is that an problem?

Reply:
After thinking about this for a bit: When I wrote this I was imagining blobs that we couldn't delete as permanent errors (a permission issue, for example), but I realize that's not true. We could definitely get an error from a delete that was transient. So I'll modify this.